node/src/reactor/participating.rs

```diff
@@ -664,75 +664,77 @@ impl reactor::Reactor for Reactor {
+                        maybe_step_effect_and_upcoming_era_validators
+                    {
+                        effects.extend(
                             effect_builder
-                                .announce_created_finality_signature(signature.clone())
-                                .await;
-                            // Allow a short period for peers to establish connections.  This delay
-                            // can be removed once we move to a single reactor model.
+                                .announce_commit_step_success(
+                                    current_era_id,
+                                    step_effect_and_upcoming_era_validators.step_execution_journal,
+                                )
+                                .ignore(),
+                        );
+                        effects.extend(
                             effect_builder
-                                .set_timeout(DELAY_FOR_SIGNING_IMMEDIATE_SWITCH_BLOCK)
-                                .await;
-                            let message = Message::FinalitySignature(Box::new(signature));
-                            effect_builder.broadcast_message(message).await
-                        }
-                        .ignore(),
-                    );
+                                .announce_upcoming_era_validators(
+                                    current_era_id,
+                                    step_effect_and_upcoming_era_validators.upcoming_era_validators,
+                                )
+                                .ignore(),
+                        );
+                    }
+
+                    // We're responsible for signing the new block if we're in the provided list.
+                    if validators_to_sign_immediate_switch_block.contains(&our_public_key) {
+                        let signature = FinalitySignature::new(
+                            *block.hash(),
+                            current_era_id,
+                            &our_secret_key,
+                            our_public_key.clone(),
+                        );
+                        effects.extend(
+                            async move {
+                                effect_builder
+                                    .announce_created_finality_signature(signature.clone())
+                                    .await;
+                                // Allow a short period for peers to establish connections. This
+                                // delay can be removed once we move to a single reactor model.
+                                effect_builder
+                                    .set_timeout(DELAY_FOR_SIGNING_IMMEDIATE_SWITCH_BLOCK)
+                                    .await;
+                                let message = Message::FinalitySignature(Box::new(signature));
+                                effect_builder.broadcast_message(message).await
```

Comment:
This makes me think that we could add a "reliable broadcast" support of some sort to the networking component so that its callers/users don't have to handle fragile things like this, restarts, etc.

/cc @marc-casperlabs

Reply:
Isn't reliable broadcast usually a building block on the consensus layer, requiring a somewhat complex protocol? I'd argue that this is "higher layer" than basic networking.

Our middle-of-the-road solution for these things is usually to gossip them, which is a "best effort broadcast".